httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/SystemDefaultCredentialsProvider.java

```diff
@@ -88,35 +93,51 @@ public void setCredentials(final AuthScope authscope, final Credentials credenti
 
     private static PasswordAuthentication getSystemCreds(
             final AuthScope authscope,
-            final Authenticator.RequestorType requestorType) {
+            final Authenticator.RequestorType requestorType,
+            final HttpContext context) {
         final String hostname = authscope.getHost();
         final int port = authscope.getPort();
         final HttpHost origin = authscope.getOrigin();
         final String protocol = origin != null ? origin.getSchemeName() :
                 (port == 443 ? "https" : "http");
+
+        final URL targetHostURL;
+        if (context != null) {
+            final HttpUriRequest httpUriRequest = (HttpUriRequest) context.getAttribute(HttpCoreContext.HTTP_REQUEST);
```

Comment:
One cannot assume the request object in the context to be always of type of HttpUriRequest. It would be safer to use the generic HttpRequest type here. 

Otherwise everything looks good to me.